test/services/network/tracing/connectivity/test_connectivity_trace.py

```diff
@@ -0,0 +1,151 @@
+#  Copyright 2022 Zeppelin Bend Pty Ltd
+#
+#  This Source Code Form is subject to the terms of the Mozilla Public
+#  License, v. 2.0. If a copy of the MPL was not distributed with this
+#  file, You can obtain one at https://mozilla.org/MPL/2.0/.
+from collections import Counter
+
+import pytest
+
+from zepben.evolve import ConnectivityResult, BasicTraversal, connected_equipment, TestNetworkBuilder, connectivity_trace, connectivity_breadth_trace, \
+    current_connectivity_trace, normal_connectivity_trace, AcLineSegment, Terminal, BusbarSection, NetworkService
+
+
+class TestConnectivityTrace:
+    network = (TestNetworkBuilder()
+               .from_junction()
+               .to_breaker(is_normally_open=True, is_open=True)
+               .to_breaker(is_normally_open=True, is_open=False)
+               .to_junction()
+               .to_breaker(is_normally_open=False, is_open=True)
+               .to_breaker(is_normally_open=True, is_open=True)
+               .to_junction()
+               .network)
+    """
+    j0--b1--b2--j3--b4--b5--j6
+        bo  no      co  bo
+
+    bo = both open
+    no = normally open
+    co = currently open
+    """
+
+    @pytest.mark.asyncio
+    async def test_connectivity_trace_ignores_open_state(self):
+        await self._validate_run(connectivity_trace(), "b2", "b1", "j0", "b4", "b5", "j6")
+        await self._validate_run(connectivity_breadth_trace(), "b2", "b1", "j0", "b4", "b5", "j6")
+
+    @pytest.mark.asyncio
+    async def test_normal_connected_equipment_trace_uses_open_state(self):
+        await self._validate_run(normal_connectivity_trace(), "b2", "b4", "b5")
+
+    @pytest.mark.asyncio
+    async def test_current_connectivity_trace_uses_open_state(self):
+        await self._validate_run(current_connectivity_trace(), "b2", "b1", "b4")
+
+    @pytest.mark.asyncio
+    async def test_doesnt_back_trace_busbars(self):
+        #
+        # ---- | ---- ----
+        #  c1 bb1 c2   c3
+        #
+        c1 = AcLineSegment(mrid="c1", terminals=[Terminal()])
+        c2 = AcLineSegment(mrid="c2", terminals=[Terminal(), Terminal()])
+        c3 = AcLineSegment(mrid="c3", terminals=[Terminal()])
+        bb1 = BusbarSection(mrid="bb1", terminals=[Terminal()])
+
+        bb_network = NetworkService()
+        bb_network.connect_terminals(bb1.get_terminal_by_sn(1), c1.get_terminal_by_sn(1))
+        bb_network.connect_terminals(bb1.get_terminal_by_sn(1), c2.get_terminal_by_sn(1))
+        bb_network.connect_terminals(c2.get_terminal_by_sn(2), c3.get_terminal_by_sn(1))
+
+        t = connectivity_trace()
+        t.process_queue.put(ConnectivityResult(c1.get_terminal_by_sn(1), bb1.get_terminal_by_sn(1), []))
+
+        visited = set()
+
+        async def step_action(cr: ConnectivityResult, _: bool):
+            visited.add(cr.to_equip.mrid)
+
+        await t.add_step_action(step_action).run()
+        assert Counter(visited) == Counter([bb1.mrid, c2.mrid, c3.mrid])
```

Comment:
question: Any reason you are checking these via `Counter` when they are sets?

You could be using `visited == {c2.mrid, bb1.mrid}` in this case. `Counter` has been used in other tests for checking that lists contain the same items without requiring the same order, but a sets `==` will do this for you.

Same applies to other uses of `Counter` on sets in the other tests.

Reply:
No reason now that you point it out. Replaced with `abc == {...}` for all set equality checks.